src/util/shm/mpir_shm_coll.c

```diff
@@ -3,36 +3,156 @@
  *  (C) 2019 by Argonne National Laboratory.
  *      See COPYRIGHT in top-level directory.
  */
+#include <mpidimpl.h>
 #include "mpir_shm.h"
 #include "mpir_shm_coll.h"
+#include "mpir_mem.h"
+#include "build_nodemap.h"
+
+#define PROC_MEM_LEN (MPIR_SHM_CACHE_LINE_LEN * 16)
+
+static int local_size;
+static int local_rank;
+static MPIR_shm_seg_t memory;
+static MPIR_shm_barrier_t *barrier;
+static void *baseaddr;
 
 int MPIR_SHM_coll_init(int rank, int size, int *nodemap)
 {
     int mpi_errno = MPI_SUCCESS;
+    int local_leader;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_SHM_COLL_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_SHM_COLL_INIT);
+
+    MPIR_NODEMAP_get_local_info(rank, size, nodemap, &local_size, &local_rank, &local_leader);
+
+    mpi_errno = MPIR_shm_seg_alloc(local_size * PROC_MEM_LEN, (void **) &baseaddr, MPL_MEM_SHM);
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    mpi_errno =
+        MPIR_shm_seg_commit(&memory, &barrier, local_size, local_rank, local_leader, rank,
+                            MPL_MEM_SHM);
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    mpi_errno = MPIR_shm_barrier(barrier, local_size);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIR_SHM_COLL_INIT);
     return mpi_errno;
+  fn_fail:
+    goto fn_exit;
 }
 
 int MPIR_SHM_coll_finalize(void)
 {
     int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_SHM_COLL_FINALIZE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_SHM_COLL_FINALIZE);
+
+    mpi_errno = MPIR_shm_barrier(barrier, local_size);
+    if (mpi_errno)
+        MPIR_ERR_POP(mpi_errno);
+
+    mpi_errno = MPIR_shm_seg_destroy(&memory, local_size);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIR_SHM_COLL_FINALIZE);
     return mpi_errno;
+  fn_fail:
+    goto fn_exit;
 }
 
 int MPIR_SHM_bcast(void *buffer, int count, MPI_Datatype datatype, int root)
 {
     int mpi_errno = MPI_SUCCESS;
+    MPI_Aint size, len;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_SHM_BCAST);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_SHM_BCAST);
+
+    /* We only support builtin datatypes */
+    MPIR_Assert(HANDLE_GET_KIND(datatype) == HANDLE_KIND_BUILTIN);
+
+    MPIR_Datatype_get_size_macro(datatype, size);
+    len = size * count;
+    MPIR_Assert(len <= PROC_MEM_LEN);
+
+    if (local_rank == root) {
+        MPIR_Memcpy(baseaddr + root * PROC_MEM_LEN, buffer, len);
+        mpi_errno = MPIR_shm_barrier(barrier, local_size);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+    } else {
+        mpi_errno = MPIR_shm_barrier(barrier, local_size);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+        MPIR_Memcpy(buffer, baseaddr + root * PROC_MEM_LEN, len);
+    }
+
+    mpi_errno = MPIR_shm_barrier(barrier, local_size);
```

Comment:
@gcongiu Why is a `barrier` needed here?

Reply:
@jain-surabhi-23 good point, not needed.